core/src/main/java/org/elasticsearch/action/index/IndexRequest.java

```diff
@@ -549,8 +558,17 @@ public void writeTo(StreamOutput out) throws IOException {
     public String toString() {
         String sSource = "_na_";
         try {
-            sSource = XContentHelper.convertToJson(source, false);
-        } catch (Exception e) {
+            if (source.length() > MAX_SOURCE_LENGTH_IN_TOSTRING) {
+                // We might slice in the middle of a UTF-8 character here, but BytesRef.utf8ToString() will read more bytes if needed.
+                // There is no risk of ArrayIndexOutOfBoundsException either as source is assumed to be properly UTF-8 encoded.
+                sSource = source.slice(0, MAX_SOURCE_LENGTH_IN_TOSTRING).utf8ToString() + "..." +
```

Comment:
You are right. It was shortsighted of me to assume `IndexRequest` would always have source in JSON. I'll refactor the code to take care of other formats too. This might take a few days though as I don't have much spare time at the moment.

P.S. I have problem setting up working environment with IntelliJ IDEA. Suggested `gradle idea` + 'Create Project from Sources' approach does not quite work (e.g. the module lacks any dependencies). Is there more up to date guide? Or would anyone be willing to share the ready-made IDEA project?